### ISSUE TITLE
Specification-based Testing walking example is wrong

### ISSUE THREAD

**Issue by thijsnulle** [sttp-book/sttp-book#189]
The only edge case that is given is where small, big and total are all -1. This is wrong because if one of the parameters is -1 the method should still pass but it fails. For example small = 20, big = -1 and total = 20 should pass but because of the line `total -= (bigBoxesWeCanUse * 5);` the total becomes 25 and the test fails. I think there should be an edge case added into the method to check for negative inputs and either put them to 0 or handle the edge cases correctly.

**Comment by mauricioaniche** [comment]
Hi, @thijsnulle, that's a good thinking!

I would not call this one a "boundary", though, but an "invalid partition". From the requirements themselves, it's clear that it makes no sense for any of those numbers to be negative. (And thinking even more, the total should be > 0, otherwise it makes no sense to mount an empty box, right?)

So, I'd definitely add "negative number" to the "small", "big", and "total" input variables. This negative number seems like an exception case, so I wouldn't really exhaustively test it in combination of other partitions to reduce the number of test cases. Maybe one test case per input would be enough.

Of course, you can see it as boundary, e.g., from when it moves from 0 to -1. That's more a matter of taste, and as long as you see, you are fine. But maybe it's easier to see that by "shifting your mind to invalid inputs". And, please, keep thinking of such exceptional cases!

If you want to open a PR (and get some bonus points), I'd suggest:

* At the end of the example in the specification-based chapter, add a remark there saying that we should also explore invalid inputs, such as -1.
* In the domain testing appendix, this is one of the examples there. I see that the -1 is not there too, so add this exceptional test there!

Cheers!

**Comment by sf105** [comment]
An interesting point to raise. It depends on where the boundaries of the unit are. Can we assume that validation is done elsewhere so we can rely on the values? If so, perhaps they should be domain types. If not, do we have to repeatedly check validity _everywhere_ in the code (as we used to in C++)?

**Comment by mauricioaniche** [comment]
We provide some examples (in the appendix) where the data is already _sanitised_. Hopefully, students get the idea that they don't have to "check everything, everywhere". But an explicit section in the book about it is _fundamental_!

I'll add that to my todo list and I'll get back there!

**Comment by BlackLeg15** [comment]
@mauricioaniche 
Shouldn't this condition be just "small < total" instead of "small <= total"?
![image](https://user-images.githubusercontent.com/32530299/106613167-aaa62d00-6548-11eb-97ee-2910458fae1a.png)

**Comment by mauricioaniche** [comment]
Good catch.... But this is on purpose! We catch this bug after discussing what _boundary testing_ is in the next chapter! :)

**Comment by BlackLeg15** [comment]
I though about that possibility, haha 
Good job, sir
